pkg/docker/config/config.go

```diff
@@ -649,30 +649,43 @@ func findAuthentication(ref reference.Named, registry, path string, legacyFormat
 		keys = []string{registry}
 	}
 
+	// It may be possible that auth.json has been externally modified to
+	// contain an entry like "https://quay.io/v1/". In that case we still want
+	// to match.
+	strippedAuths := map[string]dockerAuthConfig{}
+	for k, v := range auths.AuthConfigs {
+		strippedAuths[stripRegistry(k)] = v
+	}
+
 	// Repo or namespace keys are only supported as exact matches. For registry
 	// keys we prefer exact matches as well.
 	for _, key := range keys {
 		if val, exists := auths.AuthConfigs[key]; exists {
 			return decodeDockerAuth(val)
 		}
+		if val, exists := strippedAuths[key]; exists {
+			return decodeDockerAuth(val)
+		}
 	}
 
-	// bad luck; let's normalize the entries first
-	// This primarily happens for legacyFormat, which for a time used API URLs
-	// (http[s:]//…/v1/) as keys.
-	// Secondarily, (docker login) accepted URLs with no normalization for
-	// several years, and matched registry hostnames against that, so support
-	// those entries even in non-legacyFormat ~/.docker/config.json.
-	// The docker.io registry still uses the /v1/ key with a special host name,
-	// so account for that as well.
-	registry = normalizeRegistry(registry)
-	normalizedAuths := map[string]dockerAuthConfig{}
-	for k, v := range auths.AuthConfigs {
-		normalizedAuths[normalizeRegistry(k)] = v
-	}
+	if legacyFormat || ref == nil {
+		// bad luck; let's normalize the entries first
+		// This primarily happens for legacyFormat, which for a time used API URLs
+		// (http[s:]//…/v1/) as keys.
+		// Secondarily, (docker login) accepted URLs with no normalization for
+		// several years, and matched registry hostnames against that, so support
+		// those entries even in non-legacyFormat ~/.docker/config.json.
+		// The docker.io registry still uses the /v1/ key with a special host name,
+		// so account for that as well.
+		registry = normalizeRegistry(registry)
```

Comment:
BTW I think the key we are looking up only needs to go through the (current) `convertDockerHost`, `registry` should already be a host[port], never an URL.

Reply:
Yep, changed as suggetest.